server/src/main/java/io/deephaven/server/console/completer/PythonAutoCompleteObserver.java

```diff
@@ -0,0 +1,234 @@
+
+            PyObject completer = (PyObject) scriptSession.getVariable("jedi_settings");
+            boolean canJedi = completer.callMethod("is_enabled").getBooleanValue();
+            if (!canJedi) {
+                log.trace().append("Ignoring completion request because jedi is disabled").endl();
+                // send back an empty response. We may want to include a "don't ask again" flag in the response...
+                safelyExecuteLocked(responseObserver,
+                        () -> responseObserver.onNext(AutoCompleteResponse.newBuilder().build()));
+                return;
+            }
+            final VersionedTextDocumentIdentifier doc = request.getTextDocument();
+            final Position pos = request.getPosition();
+            final long startNano = System.nanoTime();
+
+            if (log.isTraceEnabled()) {
+                String text = completer.call("get_doc", doc.getUri()).getStringValue();
+                log.trace().append("Completion version ").append(doc.getVersion())
+                        .append(" has source code:").append(text).endl();
+            }
+            final PyObject results = completer.callMethod("do_completion", doc.getUri(), doc.getVersion(),
+                    // our java is 0-indexed lines, 1-indexed chars. jedi is 1-indexed-both.
+                    // we'll keep that translation ugliness to the in-java result-processing.
+                    pos.getLine() + 1, pos.getCharacter());
+            if (!results.isList()) {
+                throw new UnsupportedOperationException("Expected list from jedi_settings.do_completion, got " + results.call("repr"));
+            }
+            log.info().append("Got ").append(results.asList().size()).append(" completions from jedi at ")
+                    .append(pos.getLine()).append(", ").append(pos.getCharacter()).endl();
+            final long nanosJedi = System.nanoTime();
+            // translate from-python list of completion results. For now, each item in the outer list is a [str, int]
+            // which contains the text of the replacement, and the column where is should be inserted.
+            List<CompletionItem> finalItems = new ArrayList<>();
+
+            for (PyObject result : results.asList()) {
+                if (!result.isList()) {
+                    throw new UnsupportedOperationException("Expected list-of-lists from jedi_settings.do_completion, " +
+                            "got bad result " + result.call("repr") + " from full results: " + results.call("repr"));
+                }
+                // we expect [ "completion text", start_column ] as our result.
+                // in the future we may want to get more interesting info from jedi to pass back to client
+                final List<PyObject> items = result.asList();
+                String completionName = items.get(0).getStringValue();
+                int start = items.get(1).getIntValue();
```

Comment:
Hmm. I think we're still paying for each GIL trip, to do the getStringValue, getIntValue calls. Plus, each List.get call is a python `__getitem__` invoke.

I still think we can punt on this, but will want to revisit

Reply:
strange that it wouldn't just say "sending a list of primitives, lets do this all at once".

Maybe if I sent two lists, one of string, one of int, I could convince jpy to give me String[], int[] in two shots...
But, yes... for another day. I'll open some issues to track next steps.

Reply:
I've confirmed the timing is as you suspect, and will open a ticket w/ an idea how to improve it.